tensorpipe/channel/cuda_basic/channel_impl.cc

```diff
@@ -118,67 +195,124 @@ void ChannelImpl::recvImplFromLoop(
+          }));
+}
+
+void ChannelImpl::onRecvOpReadyForRecv(Operation& op, std::string descriptor) {
+  if (error_) {
+    op.done = true;
+    onRecvOpDone();
+    return;
+  }
+
+  TP_VLOG(6) << "Channel " << id_ << " is sending chunk #" << op.chunkId
+             << " of " << op.numChunks << " for buffer #" << op.sequenceNumber
              << " through CPU channel";
   cpuChannel_->recv(
       std::move(descriptor),
-      cpuBuffer,
-      callbackWrapper_([sequenceNumber,
-                        buffer,
-                        deviceIdx,
-                        tmpBuffer{std::move(tmpBuffer)},
-                        callback{std::move(callback)}](
-                           ChannelImpl& impl) mutable {
-        TP_VLOG(5) << "Channel " << impl.id_ << " is done receiving buffer #"
-                   << sequenceNumber << " through CPU channel";
-        impl.onCpuChannelRecv(
-            sequenceNumber,
-            buffer,
-            deviceIdx,
-            std::move(tmpBuffer),
-            std::move(callback));
+      CpuBuffer{op.tmpBuffer.get(), op.length},
+      callbackWrapper_([&op](ChannelImpl& impl) {
+        TP_VLOG(6) << "Channel " << impl.id_ << " is done sending chunk #"
+                   << op.chunkId << " of " << op.numChunks << " for buffer #"
+                   << op.sequenceNumber << " through CPU channel";
+        impl.onRecvOpReadyForCopy(op);
       }));
 }
 
-void ChannelImpl::onCpuChannelRecv(
-    uint64_t sequenceNumber,
-    CudaBuffer buffer,
-    int deviceIdx,
-    CudaPinnedBuffer tmpBuffer,
-    TRecvCallback callback) {
+void ChannelImpl::onRecvOpReadyForCopy(Operation& op) {
   if (error_) {
-    callback(error_);
+    op.done = true;
+    onRecvOpDone();
     return;
   }
 
-  TP_VLOG(5) << "Channel " << id_ << " is copying buffer #" << sequenceNumber
+  TP_VLOG(5) << "Channel " << id_ << " is copying chunk #" << op.chunkId
+             << " of " << op.numChunks << " for buffer #" << op.sequenceNumber
              << " from CPU to CUDA device";
-  {
-    CudaDeviceGuard guard(deviceIdx);
-    TP_CUDA_CHECK(cudaMemcpyAsync(
-        buffer.ptr,
-        tmpBuffer.get(),
-        buffer.length,
-        cudaMemcpyHostToDevice,
-        buffer.stream));
-  }
-
-  // Keep tmpBuffer alive until cudaMemcpyAsync is done.
-  cudaLoop_.addCallback(
-      cudaDeviceForPointer(context_->getCudaLib(), buffer.ptr),
-      buffer.stream,
-      callbackWrapper_([sequenceNumber, tmpBuffer{std::move(tmpBuffer)}](
-                           ChannelImpl& impl) mutable {
-        TP_VLOG(5) << "Channel " << impl.id_ << " is done copying buffer #"
-                   << sequenceNumber << " from CPU to CUDA device";
+  int deviceIdx = cudaDeviceForPointer(context_->getCudaLib(), op.cudaPtr);
+  cudaCopy(
+      op.cudaPtr,
+      op.tmpBuffer.get(),
+      op.length,
+      deviceIdx,
+      op.stream,
+      // TODO: Avoid copying op.
+      callbackWrapper_([op](ChannelImpl& impl) {
```

Comment:
Correct me if I'm wrong, but this callback is performing the crucial duty of keeping the operation alive (and thus delaying when the pinned staging buffer is returned to the allocator), right? This is far from obvious by looking at this in isolation, and based on that TODO one would be tempted to "fix" this code by avoiding to capture the op and instead capture the `chunkId`, `numChunks` and `sequenceNumber` fields separately, as these appear to be the only ones that are needed.

I think we'll fix this "automatically" with the state machine, but in the meantime it would be nice to at least add a comment about this to prevent accidental modifications.

Reply:
Added comment.